graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/processors/PipelineInterpreter.java

```diff
@@ -170,6 +186,31 @@ public Messages process(Messages messages, InterpreterListener interpreterListen
         return new MessageCollection(fullyProcessed);
     }
 
+    private void handleFailedMessage(Message message) {
+        // TODO we probably want a configuration option to disable this feature
+        // TODO if storing failed messages in ES is disabled, we don't want to mark the message as filteredOut either
+        final String processingError = message.getFieldAs(String.class, Message.FIELD_GL2_PROCESSING_ERROR);
+        if (processingError != null) {
+            message.setFilterOut(true);
+            messageQueueAcknowledger.acknowledge(message);
+            failedMessages.mark();
+            submitFailure(message, processingError);
+        }
+    }
+
+    private void submitFailure(Message message, String error) {
+        try {
+            final String messageJson = objectMapper.writeValueAsString(message.toElasticSearchObject(objectMapper, new Meter()));
```

Comment:
This is already the 2nd place where we have to manually perform the serialization before submitting a failure.

What if we in `org.graylog.failure.Failure` change the signature of the method from `String messageJson()` to `Indexable message()` and then perform all the necessary serialization in `EnterpriseFailureHandler`?

Reply:
good idea. simpler this way :+1: